### PR TITLE
First part of enforce error messages

### PR DESCRIPTION
**Pull request by alexandrasp**

### CODE REVIEW

File: jsone/builtins.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
+var {BuiltinError} = require('./error');
```
Comment (djmitche) [inline jsone/builtins.js]: Did this flie get added accidentally?

File: jsone/interpreter.py [hunk 2]
```diff
@@ -64,7 +64,7 @@ def __init__(self, context):
 
     def parse(self, expression):
         if not isinstance(expression, string):
-            raise TemplateError('expression to be evaluated must be a string')
+            raise TemplateError('expression to be evaluated must be a string, an object, a number, a boolean or null')
```
Comment (djmitche) [inline jsone/interpreter.py]: This doesn't look right....

File: jsone/render.py [hunk 3]
```diff
@@ -243,7 +243,7 @@ def reverse(template, context):
     checkUndefinedProperties(template, ['\$reverse'])
     value = renderValue(template['$reverse'], context)
     if not isinstance(value, list):
-        raise TemplateError("$reverse value must evaluate to an array")
+        raise TemplateError("$reverse value must evaluate to an array of objects")
```
Comment (djmitche) [inline jsone/render.py]: This seems incorrect, too: `{$reverse: [1, 2, 3]}` is valid, but the numbers aren't objects.

File: specification.yml [hunk 4]
```diff
@@ -928,81 +928,81 @@ title:    cannot sort objects without by
...
 template: {$sort: [1, 'a'], 'by(x)': 'x'}
-error: true
+error: 'TemplateError: $sort requires all sorted values have the same type or structure, not null and not boolean'
 ---
 title:    sort requires an array (string)
 context:  {}
 template:
   $sort:  "some string"
-error: true  # $sort must be given an array
+error: 'TemplateError: $sort value must evaluate to an array'  # $sort must be given an array
```
Comment (djmitche) [inline specification.yml]: you can remove the comment on this line and the next few copies of it

**Comment by djmitche** [comment]
@alexandrasp do you want to finish this up?  It's at risk of conflicting with other patches to the error messages..

**Comment by alexandrasp** [comment]
yes @djmitche, I want finish.

File: jsone/builtins.js (hunk 1 above)
Comment (alexandrasp) [reply jsone/builtins.js]: yes, probably

**Comment by djmitche** [comment]
Is this ready for another look?  I see that `jsone/builtins.js` still exists..

**Comment by djmitche** [comment]
Also, you might want to rebase onto master..

**Comment by djmitche** [comment]
Looks like this is failing some tests -- https://public-artifacts.taskcluster.net/EYPzPND_SgOqi2JLyGYMnA/0/public/logs/live_backing.log

**Pull request merged**